### DISCUSSION TITLE
CORS Error in Android Studio Debugging

### DISCUSSION THREAD

**Discussion by selfee-jp** [quasarframework/quasar#9423] [CLI - Capacitor / Cordova modes] [answered]
I'm getting a CORS error when debugging Android Studio.
How do you guys solve this?
The backend server is PHP.
I've looked at Capacitor and Android Studio sites, but I can't solve this problem.

Command: 
`quasar dev -m capacitor -T android`

[capacitor.config.json]
"server": {
    "hostname": "domein.net",
    "androidScheme": "https",
  }

[quasar.conf.js]
devServer: {
      https: false,
      port: true,
      open: true,
      allowedHosts: [
        'domein.net'
      ],
},
capacitor: {
　  hideLogs:false,
      hideSplashscreen: true,
      "plugins": {
        "PushNotifications": {
          "presentationOptions": ["badge", "sound", "alert"]
        }
      },
"server":{
        "hostname": "accentstyle.xsrv.jp",
        "iosScheme": "https",
        "androidScheme": "https"
}

[server PHP]
header("Access-Control-Allow-Origin: *");
header('Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept');
header('Content-Type: application/json; charset=utf-8');
header("Access-Control-Allow-Credentials:true");

**Comment by hawkeye64** [comment]
@selfee-jp CORS is a function of the server, not the client. 
https://enable-cors.org/server_php.html

**Reply by selfee-jp** [reply]
@hawkeye64  I understand about that too.
I'm able to test it at https://www.test-cors.org/.
The result is fine.

But why can't I debug it from "Android Studio"?
Also, I can't communicate with it even after building, signing, and installing it.

**Reply by ontwikkelfabriek** [reply]
@selfee-jp 
> I'm getting a CORS error when debugging Android Studio.

So does that mean running the app normally doesn't give you problems/cors errors?


> I'm able to test it at https://www.test-cors.org/.

On that site you can test 2 things: 
_You can either send the CORS request to a remote server (to test if CORS is supported), or send the CORS request to a test server (to explore certain features of CORS)._ 

What did you test exactly? ( you should test both btw)


About your config:

>_[quasar.conf.js]
devServer: {
**https: false,**
**port: true,**
open: true,
**allowedHosts: [
'domein.net'
]**


1) https: **true**
2) port: 8080 or 8088 -> a Number **not** a Boolean
3) allowedHosts -> remove 

>"server":{
"hostname": "accentstyle.xsrv.jp",
"iosScheme": "https",
"androidScheme": "https"
}

What/where is this config block for?

 > But why can't I debug it from "Android Studio"?

Why would you want to debug a Quasar app from Android studio anyways, instead of Chrome debugger using `chrome://inspect/#devices`?

> Also, I can't communicate with it even after building, signing, and installing it.

I do not think you can attach a debugger to android apps that are signed, try just building the app instead without signing.

Please post the output of you cors error.

To bypass cors issues for api request to (remote) servers in dev mode you can use this:
https://quasar.dev/quasar-cli/api-proxying#introduction

**Reply by selfee-jp** [reply]
@ontwikkelfabriek 
Maybe I'm not understanding.
Please help me.

> So does that mean running the app normally doesn't give you problems/cors errors?

No, it's not working.

> On that site you can test 2 things:
You can either send the CORS request to a remote server (to test if CORS is supported), or send the CORS request to a test server (to explore certain features of CORS).

Result.
Fired XHR event: loadstart
Fired XHR event: readystatechange
Fired XHR event: readystatechange
Fired XHR event: progress
Fired XHR event: readystatechange
Fired XHR event: load

XHR status: 200
XHR status text:
XHR exposed response headers:
cache-control: no-store, no-cache, must-revalidate
content-type: text/html;charset=UTF-8
expires: Thu, 19 Nov 1981 08:52:00 GMT
pragma: no-cache

> What/where is this config block for?

I was looking at a lot of information and trying it out.
Therefore, there may be some blocks that are not needed in Quasar.

[quasar.conf.js]
```
devServer: {
      https: false,
      port: 8080,
      open: true // opens browser window automatically
},
capacitor: {
      hideSplashscreen: true
},
```

> Why would you want to debug a Quasar app from Android studio anyways, instead of Chrome debugger using chrome://inspect/#devices?

This is to create a push notification application.
We believe that the actual device is not communicating with us either.

> I do not think you can attach a debugger to android apps that are signed, try just building the app instead without signing.

I could not install it on the actual device without signing it. Normally, you should be able to install without a signature. I am considering this issue separately.

> Please post the output of you cors error.

I reworked the configuration in Quasar.
The result will not change.

I/Capacitor/Console: File: webpack-internal:///./src/http.js - Line 39 - Msg: POST https://example.jp/example/login >> {"login":{"tok":"","id":0},"data":{"account":"1","password":"1"}}

E/Capacitor/Console: File: http://192.168.0.2:8080/#/ - Line 0 - Msg: Access to XMLHttpRequest at 'https://example.jp/example/login' from origin 'http://192.168.0.2:8080' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.

E/Capacitor/Console: File: webpack-internal:///./src/http.js - Line 22 - Msg: Uncaught (in promise) Error: API error.

-----
"boot" specifies the URL of the API.
`Vue.prototype.$BasePostUrl = "https://example.jp/example/mobile";`

"https.js"

```
const debug = process.env.NODE_ENV !== 'production'
const onSuccess = (resp) => {
  if (debug) {
    console.log(' << ' + JSON.stringify(resp.data))
  }
  return Promise.resolve(resp.data)
}
const onError = () => {
  throw new Error('API error.')
}
export default {
  get: (url, params) => {
    if (debug) {
      console.log('GET ' + url + ' >> ' + JSON.stringify(params))
    }
    return axios.get(url, { params: params }).then(onSuccess).catch(onError)
  },
  post: (url, params) => {
    console.log(params);
    if (debug) {
      console.log('POST ' + url + ' >> ' + JSON.stringify(params))
    }
    return axios.post(url, params,{
      mode: 'no-cors',
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json',
      },
      withCredentials: true,
      //credentials: 'same-origin',
    }).then(onSuccess).catch(onError)
  },
}
```
 ----
Use "https.js" in the root store.

```
import Axios from '../http'
state:{
    baseUrl:Vue.prototype.$BasePostUrl,
  },
actions:{
      async axios_post (e,params) {
        let UserInfo = {}
        if ( e.state.auth.user !=null ){
          UserInfo = {
            tok:e.state.auth.user.tok,
            id:e.state.auth.user.id
          }
        }else{
          UserInfo = {
            tok:"",
            id:0
          }
        }  
          let post = {
            login : UserInfo,
            data : params.param
          }
          return await Axios.post(params.url,post);
      },
}
```

Called from another action.

```
await e.dispatch("axios_post",{url:e.rootState.baseUrl+"/login",param:payload},{root:true})
  .then( response => {});
```
-----
Server side.
Folder structure.
/example.jp/public_html/example/index.php

.htaccess

```
RewriteEngine On
RewriteCond %{REQUEST_FILENAME} !-f
RewriteRule ^(.*)$ index.php [QSA,L]
```

https://example.jp/example/mobile/login will be called.
```
require 'config.php';
require 'dbClass.php';
require 'Slim/Slim.php';

\Slim\Slim::registerAutoloader();
$app = new \Slim\Slim();

header("Access-Control-Allow-Origin: *");
header('Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept');
header('Content-Type: application/json; charset=utf-8');
header("Access-Control-Allow-Credentials:true");

$app->post('/login','login');
$app->run();

function login() {
	$request = \Slim\Slim::getInstance()->request();
	$data = json_decode($request->getBody());


	try {
		$res["status"] = true;
		echo json_encode($res);
	}
	catch(PDOException $e) {
		echo '{"status":false,"error":{"text":'. $e->getMessage() .'}}';
	}
}
```

**Reply by metalsadman** [reply]
@selfee-jp 
> I could not install it on the actual device without signing it. Normally, you should be able to install without a signature. I am considering this issue separately.

you can use `--debug` flag on your build command to make debug package of your app and can install it without signing.

**Reply by selfee-jp** [reply]
@metalsadman 

> you can use --debug flag on your build command to make debug package of your app and can install it without signing.

Thank you. I was able to install it unsigned.
`quasar build -m capacitor -T android --debug`

But I haven't been able to resolve the CROS error.

**Reply by selfee-jp** [reply]
@ontwikkelfabriek

> To bypass cors issues for api request to (remote) servers in dev mode you can use this:

Indeed, I was able to solve the problem by using a proxy.
But isn't this a development setting?
Do I still need a proxy for signed builds that I want to publish?

```
devServer: {
      https: false,
      port: 8080,
      open: true,
      proxy: {
        '/login': {
          target: 'https://example.jp/example/login',
          changeOrigin: true,
        }
      }
},

```

**Reply by ontwikkelfabriek** [reply]
@selfee-jp 

>_The value of the 'Access-Control-Allow-Origin' header in the response must **not** be the **wildcard** '*' when the request's credentials mode is **'include'**. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the **withCredentials** attribute._

Looking at this message and your config, I think these 2 lines conflict:

[server PHP]
--->header("Access-Control-Allow-Origin: *");
header('Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept');
header('Content-Type: application/json; charset=utf-8');
--->header("Access-Control-Allow-Credentials:true");

The error states that `Access-Control-Allow-Origin` cannot have a value wildcard(*) if you use credentials, = `Access-Control-Allow-Credentials`. I would try to set `Access-Control-Allow-Credentials:**false**` or set Access-Control-Allow-Origin: '**http://192.168.0.2:8080**' . Besides that,  because your server run on https, I would try to set devServer to **https = true**

> Indeed, I was able to solve the problem by using a proxy.
But isn't this a development setting?
Do I still need a proxy for signed builds that I want to publish?

For production you probably have to set the `Access-Control-Allow-Credentials` to `localhost:8080` instead of an IP number.

Also you'll not get cors errors in a Capacitor App if you use a Native http library to do request instead of using the Webview. You could check out this( part 2 using native http):
https://hackernoon.com/a-practical-solution-for-cors-cross-origin-resource-sharing-issues-in-ionic-3-and-cordova-2112fc282664

for Capacitor I Found this:
https://github.com/capacitor-community/http

**Reply by selfee-jp** [reply]
@ontwikkelfabriek 

> Looking at this message and your config, I think these 2 lines conflict:

I am going to delete Credentials.

[server PHP]
```
header("Access-Control-Allow-Origin: *");
header('Content-Type: application/json; charset=utf-8');
```


> Also you'll not get cors errors in a Capacitor App if you use a Native http library to do request instead of using the Webview. You could check out this( part 2 using native http):

I see that you are using Native http library instead of Axios for development with Quasar + Capasitor.
Does the lack of this topic mean that you guys don't have a problem with it?

Am I correct in assuming that you use the proxy for development and the Native http library for release?
It seems a bit time-consuming.

Also, in the product build, the proxy is working. Do you think this is a bug?

**Reply by selfee-jp** [reply]
@ontwikkelfabriek 

Unfortunately, I tried using @capacitor-community/http, but the communication failed.
`
Access to fetch at 'https://example.jp/example/login' from origin 'http://192.168.0.2:8080' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status.`

I also checked the version of @capacitor-community/http.
For Capacitor 2.x projects, you will need to install a version less than 1.0.0. You can do that by specifying the version in your package.json or installing like this. The latest 2.x compatible version is 0.3.1.
```
  "dependencies": {
    "@capacitor-community/http": "^0.3.1",
    "@capacitor/android": "^2.4.7",
    "@capacitor/cli": "^2.0.0",
    "@capacitor/core": "^2.0.0"
  }
```
[server PHP]
```
header("Access-Control-Allow-Origin: *");
header('Content-Type: application/json; charset=utf-8');
```

[front]
```
const onSuccess = (resp) => {
  return Promise.resolve(JSON.parse(resp.data))
}
const onError = () => {
  throw new Error('API error.')
}
post: (url, params) => {

  const options = {
        method: 'POST',
        url: url,
        headers: {
          'Content-Type': "application/json"
        },
        data: params,
      };
  return Http.request(options).then(onSuccess).catch(onError);
}
```

 If you do not set the 'Content-Type', it will communicate.
However, the server will not be able to retrieve the posted value.
```
const options = {
        method: 'POST',
        url: url
        data: params,
      };
```

I've decided to return 200ok for preflight requests.
I'm not sure if this is correct.

[server PHP]
```
$method = $_SERVER['REQUEST_METHOD'];
if ($method == "OPTIONS") {
	header('Access-Control-Allow-Origin: *');
	header('Content-Type: application/json; charset=utf-8');
	header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS');
	header("Access-Control-Allow-Headers: Origin, Content-Type");
	header("HTTP/1.1 200 OK");
	die();
}else{
	header("Access-Control-Allow-Origin: *");
	header('Content-Type: application/json; charset=utf-8');
	header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS');
	header('Access-Control-Allow-Headers: Origin, Content-Type');
}
```

I installed it on an actual device and checked the communication.
Success.

`quasar build -m capacitor -T android --debug`

Checked the communication in Android studio.
Success.

`quasar deve -m capacitor -T android`

Perhaps there is no problem with Axios?

**Reply by ontwikkelfabriek** [reply]
>I see that you are using Native http library instead of Axios for development with Quasar + Capacitor.
Does the lack of this topic mean that you guys don't have a problem with it?

Correct I don't have cors problems with Axios, but I have always been using Cordova. 

> Am I correct in assuming that you use the proxy for development and the Native http library for release?
It seems a bit time-consuming.

No I just pointed you to the native http library to try and see if that would solve some things. It did not so your problem is/was server side.

> Also, in the product build, the proxy is working. Do you think this is a bug?

eum I don't think the proxy is working, the proy just redirects api call's to be send from your devMachine's server instead of the client. (request from servers **don't** have cor's problems). In production your requests will go directly from the device to the remote server. 

> Perhaps there is no problem with Axios?

As mentioned before cors problems are `usually` caused by server side (mis)configuration.

**Comment by selfee-jp** [comment, marked as answer]
Thank you, @ontwikkelfabriek 
Thanks to everyone who gave me advice.

Cors problems were usually caused by (incorrect) server-side settings.
I should have noticed the preflight request.
In the preflight request, I was able to solve the problem by sending back a 200 OK.
In PHP, I was able to solve the problem as follows.

[server PHP]
```
$method = $_SERVER['REQUEST_METHOD'];
if ($method == "OPTIONS") {
	// When a preflight request is received.
	header('Access-Control-Allow-Origin: *');
	header('Content-Type: application/json; charset=utf-8');
	header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS');
	header("Access-Control-Allow-Headers: Origin, Content-Type");
	header("HTTP/1.1 200 OK");
	// Return only the header and stop the process.
	exit();
}else{
	header("Access-Control-Allow-Origin: *");
	header('Content-Type: application/json; charset=utf-8');
	header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS');
	header('Access-Control-Allow-Headers: Origin, Content-Type');
}
```

**Reply by hawkeye64** [reply]
You might also want to add 'HEAD' to your allowed methods, since this is usually part of the pre-flight request

On my Express server, I have the following (it may be useful for you to extrapolate it to PHP):

```js
  app.use(function (req, res, next) {
    res.header('Access-Control-Allow-Origin', req.get('Origin') || '*')
    res.header('Access-Control-Allow-Credentials', 'true')
    res.header('Access-Control-Allow-Methods', 'GET,HEAD,PUT,PATCH,POST,DELETE')
    res.header('Access-Control-Expose-Headers', 'Content-Length')
    res.header('Access-Control-Allow-Headers', 'Accept, Authorization, Content-Type, X-Requested-With, Range')
    return next()
  })
```

**Reply by selfee-jp** [reply]
I am using Slim2.
The latest Slim4 seems to be able to solve this problem.

For each URL that is requested.

```
$app->options('/', function (Request $request, Response $response): Response {
    return $response;
});
```
